File: aoc/2024/aoc202404.hs
module AOC202404 where

import Data.Array(Array,bounds,inRange,range,(!))

import AOC

aoc = AOC {
    day="04",
    aocTests=[
        AOCTest {
            testData=unlines [
                "MMMSXXMASM",
                "MSAMXMSMSA",
                "AMXSXMAAMM",
                "MSAMASMSMX",
                "XMASAMXAMM",
                "XXAMMXXAMA",
                "SMSMSASXSS",
                "SAXAMASAAA",
                "MAMMMXMMMM",
                "MXMXAXMASX"
                ],
            testResult=Just "18",
            testResult2=Just "9"
            }
        ],
    aocCode=Code {
        codeParse=parse2da,
        codeParse2=parse2da,
        codeTest=result,
        codeTest2=result2,
        codeResult=result,
        codeResult2=result2
        }
    }

result :: Array (Int,Int) Char -> Int
result grid = sum $ map count $ range $ bounds grid
  where
    count (x,y)
      | grid!(x,y) /= 'X' = 0
      | otherwise = sum [mas dxdy |
            dxdy <- [(1,0),(1,1),(1,-1),(0,1),(0,-1),(-1,0),(-1,1),(-1,-1)]]
      where
        mas (dx,dy)
          | not (inRange (bounds grid) (x+3*dx,y+3*dy)) = 0
          | grid!(x+dx,y+dy) /= 'M' = 0
          | grid!(x+2*dx,y+2*dy) /= 'A' = 0
          | grid!(x+3*dx,y+3*dy) /= 'S' = 0
          | otherwise = 1

result2 grid = length $ filter xmas $ range $ bounds grid
  where
    xmas (x,y)
      | grid!(x,y) /= 'A' = False
      | not (inRange (bounds grid) (x+1,y+1)) = False
      | not (inRange (bounds grid) (x-1,y-1)) = False
      | notMS (grid!(x+1,y+1),grid!(x-1,y-1)) = False
      | notMS (grid!(x-1,y+1),grid!(x+1,y-1)) = False
      | otherwise = True
    notMS ('M','S') = False
    notMS ('S','M') = False
    notMS _ = True

File: aoc/2025/aoc202404.go
package main

func init() {
	Register(&aoc202404{
		AOC: AOC{
			Day:           4,
			InputFilename: "../2024/input/04.txt",
			Tests: []Test{
				Test{
					Input: `MMMSXXMASM
MSAMXMSMSA
AMXSXMAAMM
MSAMASMSMX
XMASAMXAMM
XXAMMXXAMA
SMSMSASXSS
SAXAMASAAA
MAMMMXMMMM
MXMXAXMASX
`,
					Part1: "18",
					Part2: "9",
				},
			},
		},
	})
}

type aoc202404 struct {
	AOC
}

func (aoc *aoc202404) Part1(input *Input) string {
	_, _, grid := input.Grid()
	result := 0
	for xy, ch := range grid {
		if ch != 'X' {
			continue
		}
		x := xy[0]
		y := xy[1]
		for _, dx := range [3]int{-1, 0, 1} {
			for _, dy := range [3]int{-1, 0, 1} {
				if grid[[2]int{x + dx, y + dy}] == 'M' && grid[[2]int{x + 2*dx, y + 2*dy}] == 'A' && grid[[2]int{x + 3*dx, y + 3*dy}] == 'S' {
					result++
				}
			}
		}
	}
	return IntResult(result)
}

func (aoc aoc202404) mas(a, b byte) bool {
	return (a == 'M' && b == 'S') || (a == 'S' && b == 'M')
}

func (aoc aoc202404) Part2(input *Input) string {
	_, _, grid := input.Grid()
	result := 0
	for xy, ch := range grid {
		if ch != 'A' {
			continue
		}
		x := xy[0]
		y := xy[1]
		if aoc.mas(grid[[2]int{x - 1, y - 1}], grid[[2]int{x + 1, y + 1}]) && aoc.mas(grid[[2]int{x + 1, y - 1}], grid[[2]int{x - 1, y + 1}]) {
			result++
		}
	}
	return IntResult(result)
}
